### PR TITLE
Reduce memory allocations in network layer (3d edition)

### PR DESCRIPTION
**Pull request by amoerie**
The next batch of performance / memory improvements. This will focus on DicomService and PDataTFStream.

1st edition: #1267
2nd edition: #1273

Related issue: #1262 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Add dependency on Microsoft.Toolkit.HighPerformance.Buffers and use MemoryOwner
- Use rented byte arrays when buffering incoming raw PDUs (implementation detail: only when PDU < 1 MB)
- Avoid intermediate buffer copy in DicomService
- Use rented byte arrays inside PDataTFStream when writing P-DATA-TF
- Make separate classes for dependencies of DicomServer and DicomService so we can change these dependencies later without breaking consumers
- Introduce new abstraction: IMemoryProvider and replace all usages of ArrayPool<byte>.Shared with this new IMemoryProvider

### CODE REVIEW

File: FO-DICOM.Core/Network/DicomService.cs [hunk 1]
```diff
@@ -1715,6 +1752,19 @@ private async Task WritePDUAsync(bool last)
                     _pdu = new PDataTF();
                 }
             }
+            
+            private byte[] RentNextByteArray(int length)
```
Comment (gofal) [inline FO-DICOM.Core/Network/DicomService.cs]: maybe other naming like "RentEvenLengthArray(int minLength)", When reading the method I associated it like there there was a linked chain of arrays and this method will rent the next one.

**Comment by gofal** [comment]
I see that we are now using ArrayPool, because of reusing rented arrays and therefore reducing memory allocations. But this now changes code pretty much, because we have to take care about returning the rented array in finalizers, and the rented array may be larger then the requested length and therefore we have to store the requested length in separate fields.
When reviewing the code, I thought, this must be simpler, by wrapping the rented array into a IDisposable class. And actually, Microsoft did that already for us:
https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/memoryowner
Have you already taken a look at that class? using the MemoryOwner<T> instead of rented arrays may make have the same memory improvement, but make the code better readable again - and since the rented memory is wrapped, it is more save to avoid memory leaks in case we forget to release the rented array somewhere.

**Comment by amoerie** [comment]
@gofal 

I agree that bare ArrayPool usage is a little "low level", and it has pitfalls, but for cases where the array is only used locally and is very short lived (like in most usages inside the PDU file) I think it is fine as the entire lifetime of the array is contained inside a try { ... } finally { ... } block where the rented array is returned inside the finally block. 

However, if the lifetime of the array becomes more vague, I agree that some sort of abstraction would be very useful so we can implement safety nets using finalizers. I haven't used MemoryOwner yet, but I agree that it looks like an ideal solution for how we're using ArrayPool. Is it okay to add this new dependency then? Looks like it supports .NET Standard 2.0, so we definitely could.

Furthermore, I wanted  to ask you whether it might be a good idea to make a "MemoryManager", much in the same way that we already have a NetworkManager, LogManager, etc. today. This would allow expert consumers of fo-dicom to swap in their own memory allocation implementation. ( Perhaps something based on https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream ) 
We could provide a default implementation using MemoryOwner from the high performance toolkit. 

I imagine the interface would look a little like this:

```csharp
public interface IMemoryManager {
  IMemory GetMemory(int minimumLength);
}

public interface IMemory : IDisposable {
  byte[] Bytes { get; }
  int MinimumLength { get; }
}
```
(Note: Microsoft is pushing towards using Span<byte> and Memory<byte> instead of byte[], but I'm not confident all of our usages would be compatible)

To be clear, I would not always use this memory manager. ArrayPool<byte>.Shared is highly optimized and safe to always use. 
However, its design is optimized for small (< 1MB), short-lived arrays. Short-lived means things like intermediate buffers mostly. 

In other words, I would still prefer to use ArrayPool<byte>.Shared in cases where the buffer fits these criteria perfectly. Examples for fo-dicom would include things like PDU where we write an association request.

However, the MemoryManager would be preferred if the lifetime can get more complex and the size can get bigger. Examples for fo-dicom would include things like PDataTFStream and incoming Raw PDUs in DicomService.
Further down the road, we could look at image rendering and other places too, but that might require public signature changes.

File: FO-DICOM.Core/Network/DicomService.cs (hunk 1 above)
Comment (amoerie) [reply FO-DICOM.Core/Network/DicomService.cs]: Ah I understand the confusion. I will change the name!

**Comment by gofal** [comment]
A Reference to Microsofts HighPerformance package is fine. This package is supported from netstandard1.4 upwards. 

Sure, when renting a 6 byte array for the preamble, then the overhead of a wrapping MemoryOwner would be too much.

But I even would prefer using the MemoryOwner, or some similar wrapper over the rented arrays from ArrayPools in cases where both would fit and the overhead of the MemoryOwner is not breaking performance. Because the simple fact that the rented array may be larger than the requested size and you have to store somewhere the actual length is worth the wrapper and helps preventing future bugs.

Having the MemoryManager would be fine, if we really everywhere would use an Interface instead of byte arrays. Eg if we would use Streams always instead of bytes arrays. Or Span, which will be available once we would drop support for netframework. So I am not sure if we can implement that in an so abstract way, that injecting a own implementation is bringing advantages. The NetworkManager is only implemented, because in test-cases there is a mock for slow network. I doubt that someone really implents his own networkmanager instance.

The Manager-classes are kind of deprecated and should be replaced by the DI-Container. So there is absolute nothing to say against registering the MemoryStream as IMemoryStream and instead of creating a new instance of MemoryStream, requesting a service from the serviceprovider. Then a user can register his own implementation of IMemoryStream, like the RecycalbeMemoryStream.

**Comment by gofal** [comment]
You are now optimizing Memory in netwok communication. 
Optimizing memory when it comes to loading/storing, there are many temporary streams, most of them MemoryStreams, that internally always allocate new byte arrays. Or when Decoding is done, then many parts of the memory are copied. So there will be some place for optimization. 
I do have an issue, that writing performance of WSI files (some gigabytes) is terrible, because everything is kept in memory. i will have to do quite some optimizations there, but i don't want to have code conflicts, thats why i am waiting for you to be done with that all, before I start.

**Comment by amoerie** [comment]
> A Reference to Microsofts HighPerformance package is fine. This package is supported from netstandard1.4 upwards.
> 
> Sure, when renting a 6 byte array for the preamble, then the overhead of a wrapping MemoryOwner would be too much.
> 
> But I even would prefer using the MemoryOwner, or some similar wrapper over the rented arrays from ArrayPools in cases where both would fit and the overhead of the MemoryOwner is not breaking performance. Because the simple fact that the rented array may be larger than the requested size and you have to store somewhere the actual length is worth the wrapper and helps preventing future bugs.

Alright, I suppose the miniscule overhead of MemoryOwner is worth the added safety net and developer productivity in the form of code readability.

> Having the MemoryManager would be fine, if we really everywhere would use an Interface instead of byte arrays. Eg if we would use Streams always instead of bytes arrays. Or Span, which will be available once we would drop support for netframework. So I am not sure if we can implement that in an so abstract way, that injecting a own implementation is bringing advantages. The NetworkManager is only implemented, because in test-cases there is a mock for slow network. I doubt that someone really implents his own networkmanager instance.
> The Manager-classes are kind of deprecated and should be replaced by the DI-Container. 

Technically you are definitely correct, anyone can override interface implementations. What I like about the "Manager" style though is that it makes it explicit that this is an "extension point", where fo-dicom explicitly allows custom implementations. 
I agree that DI is perfect for "overriding an implementation", but I would prefer to be careful and deliberate in which implementations are "supported to be overridden". At the very least, this is something we should document. Preferably it should be obvious in the code, e.g. via extension methods on IServiceCollection like we have now with the "Manager" implementations. 

> So there is absolute nothing to say against registering the MemoryStream as IMemoryStream and instead of creating a new instance of MemoryStream, requesting a service from the serviceprovider. Then a user can register his own implementation of IMemoryStream, like the RecycalbeMemoryStream.

So I'll take it as a "no" on the MemoryManager question, but "yes" on some simple interface? What if we just rename it then?

Something like 

```csharp
public interface IMemoryProvider {
  IMemory Provide(int minimumLength);
}

public interface IMemory : IDisposable {
  byte[] Bytes { get; }
  int MinimumLength { get; }
}
```

In the future we can modify the IMemory interface to make use of more performant APIs such as Span, Memory, BufferWriter, etc. This will require investigation and benchmarking, so for now I would keep it simple with byte[].

> Optimizing memory when it comes to loading/storing, there are many temporary streams, most of them MemoryStreams, that internally always allocate new byte arrays. Or when Decoding is done, then many parts of the memory are copied. So there will be some place for optimization.

Yes, I believe we can make fo-dicom a lot faster still. Not only the byte copying when reading and writing files, but I also think DicomService could be a lot faster and more optimized if we reduced the amount of locks needed, changed the way the internal message and PDU queueing works, and used something like System.IO.Pipes for the incoming and outgoing communication.
Furthermore, each DicomItem in a DicomDataset gets his own IByteBuffer allocated. It might be a lot more efficient to have one ReadonlySequence of byte to represent the Dicom file, and then give every DicomItem a ReadonlyMemory of byte that represents its contents.
However, all of this would only work if we introduce the concept of a "ReadonlyDicomDataset" which could be a lot more performant and simpler because it does not have to be writable.

So many ideas, so little time. :-)

**Comment by gofal** [comment]
> So many ideas, so little time. :-)
👍 🥇

**Comment by gofal** [comment]
concerning the MemoryManager or MemoryProvider. The naming does not matter. But the huge difference is: the manager is one static thing, while with DI you can configure and inject more precisely. If you are running within a console application, then the DI also breaks down into one static Setup.ServiceProvider, so there they would be the same. But as soon as you are running in asp.net or some other environment with built in DI, this gives you far more control and possiblities.

The various services, that the user can override via DI, are documented in read.me (this are only very few at the moment), and of course there are already extension methods for those.

But to really give the user the possibility to inject his own memory implementation, we would have to use it very extensively. Namely at least everywhere, where you now used ArrayPool, and everywhere, where a new MemoryStream is instanciated etc. And the interface requires quite some entry-points, like allocating, freeing, maybe locking, etc. I doubt if it's worth it, at least not in such core code like here, where you are doing optimizations anyway.

Maybe when it comes to long-lasting memory, like when reading a file, where we now use MemoryStream. Maybe someone wants to inject there and use TempFileStream for all of those to reduce the heap size dramatically.

**Comment by gofal** [comment]
> Furthermore, each DicomItem in a DicomDataset gets his own IByteBuffer allocated. It might be a lot more efficient to have one ReadonlySequence of byte to represent the Dicom file, and then give every DicomItem a ReadonlyMemory of byte that represents its contents.
However, all of this would only work if we introduce the concept of a "ReadonlyDicomDataset" which could be a lot more performant and simpler because it does not have to be writable.

some time ago, i managed to reduce memory drastically, by extending the IBytebuffer with some stream-methods. Previously, there have always been new byte arrays where data is copied many times. with the streams, now one MemoryStream is holding the actual byte-array, and on top of those there are streams only pointing on parts of this memorystream. A similar concpet as there is now with Span. But Span is more low-level, since it can also work with pinned unmanaged memory. But having streams more consequently will already improve a lot.

**Comment by mrbean-bremen** [comment]
Just wanted to say that I appreciate all the work you are doing and the mass of good ideas. I have no idea where you take the time for that...
I don't have many thoughts here other than I think that all these performance optimizations are a very good thing, and important to increase the acceptance of the framework. As for the concepts: with respect to the memory manager, I'm with @gofal here - I don't think this is needed, at least at this point in time. I have mentioned before that I would like to have some wrapper that handles the `ArrayPool` disposal for us if we call it via `using`, and `MemoryOwner` seems already to do this, so it sounds good to me. Also the ideas by @amoerie regarding the speedup of `DicomService` sound great to me.

**Comment by amoerie** [comment]
There are only a few places left where ArrayPool<byte>.Shared is used directly, I'll have to resort to Setup.ServiceProvider.GetRequiredService<IMemoryProvider> though because a lot of these are in static places.

Everything else already uses a new interface called `IMemoryProvider` which provides an instance of `IMemory` (which is `IDisposable`)
This is a thin wrapper around MemoryOwner, which leaves us the flexibility of changing the API later.

**Comment by gofal** [comment]
Thanks. the new interface is mainly easier to read and more save against memory leaks due to the IDisposable interface.
But did you notice any difference in performance/memoryconsuption? i assume it did not get better, but did it get worse now with that new tier?

File: Tests/FO-DICOM.Benchmark/Program.cs [hunk 2]
```diff
@@ -10,7 +10,7 @@ static class Program
         static void Main()
         {
             // Run all benchmarks in assembly
-            BenchmarkRunner.Run(typeof(Program).Assembly);
+            BenchmarkRunner.Run<ServerBenchmarks>();
```
Comment (gofal) [inline Tests/FO-DICOM.Benchmark/Program.cs]: unintended restricted benchmark execution only to ServerBenchamarks?

Comment (amoerie) [reply Tests/FO-DICOM.Benchmark/Program.cs]: Ah yes this is temporary. This will be a good reminder to revert that.

**Review by gofal — APPROVED** [review]

**Comment by amoerie** [comment]
> But did you notice any difference in performance/memoryconsuption? i assume it did not get better, but did it get worse now with that new tier?

I will look into this when I can, but I imagine the overhead will be very very tiny.
If we start using rented memory in the pixel rendering too, then there it might make a bigger difference if we rent buffers in a hot path. But for the network layer I think the overhead will be miniscule.

**Comment by amoerie** [comment]
There is a breaking change in this PR that we haven't talked about: I introduced DicomServiceDependencies and DicomServerDependencies. These classes replace the list of parameters in the constructors of DicomService and DicomServer respectively.
This is a breaking change, since everyone that has a DicomServer and upgrades to 5.0.3 (assuming this is in the next release) will have to modify some code to also inject DicomServiceDependencies or DicomServerDependencies in the constructors of their derived classes.

I made this change so that the IMemoryProvider can be injected from DI (this would have been a breaking change too), but instead of simply adding a new parameter, I opted for creating a new class instead. This has the benefit of being a breaking change only once. 
In the future, if we want to remove/add any dependencies from DicomServiceDependencies/DicomServerDependencies, this will no longer be a breaking change.

I was wondering if you noticed that @gofal and if that is okay with you. At the very least it should be properly documented in the release notes, but I'm not sure where I can write it down so it will be visible when you make a release.

Example:

```csharp
public DicomCEchoProvider(INetworkStream stream, Encoding fallbackEncoding, ILogger log,
            ILogManager logManager, INetworkManager networkManager, ITranscoderManager transcoderManager)
            : base(stream, fallbackEncoding, log, logManager, networkManager, transcoderManager)
```

will have to be modified to

```csharp
public DicomCEchoProvider(INetworkStream stream, Encoding fallbackEncoding, ILogger log,
            DicomServiceDependencies dependencies)
            : base(stream, fallbackEncoding, log, dependencies)
````

**Comment by gofal** [comment]
Yes, I noticed that. And I also found myself often typing many many parameters to the constructor for just passing them to the base constructor and never using them when creating a new DicomService.

This is kind of philosophy how to consume services from DI. In old times the code called the WHATEVERManager.Instance property of any service and no one using that classes knew which services are used under the hood. 
With DI this is more clear and can be injected better by registering in DI, or by directly passing mockec instances in unitttests. so a very good enhancement.
But DI not only supports constructor injection, but also propery injection. So the Activator iterates the properties of the newly created instances and initializes all properties with an instance of the serviceprovider, if the type of that property was registered.
Many dicusstion groups I read about DI came to the conclusion, that constructor injection is better than property injection, because it makes it more obvious which other services are required by a class. But the downside is a long (and maybe growing) list of parameters in a constructor.

I personally do have no preference. This new class DicomServiceDependencies feels like a kind of mixture of both approaches. I would also be fine with moving those dependening services to property injection.

**Comment by mrbean-bremen** [comment]
I like the `DicomServiceDependencies` idea, but I think the change should at least need a minor version update (e.g. 5.1). I think you were to follow semantic versioning here, but I may confuse this with another repo...

**Comment by gofal** [comment]
except version 1.1, there has never been a x.1 version in fo-dicom history. main version changed when there was a greater internal redesign and else the thrid version number just increased. Breaking changes are listed in relesenotes. having version 5.1 out is also fine for me.

**Comment by mrbean-bremen** [comment]
Ok, in this case I _did_ confuse this with something else that uses semantic versioning (I tend to use it myself, too) - you can ignore my comment then....

**Comment by amoerie** [comment]
> Yes, I noticed that. And I also found myself often typing many many parameters to the constructor for just passing them to the base constructor and never using them when creating a new DicomService.
> 
> This is kind of philosophy how to consume services from DI. In old times the code called the WHATEVERManager.Instance property of any service and no one using that classes knew which services are used under the hood. With DI this is more clear and can be injected better by registering in DI, or by directly passing mockec instances in unitttests. so a very good enhancement. But DI not only supports constructor injection, but also propery injection. So the Activator iterates the properties of the newly created instances and initializes all properties with an instance of the serviceprovider, if the type of that property was registered. Many dicusstion groups I read about DI came to the conclusion, that constructor injection is better than property injection, because it makes it more obvious which other services are required by a class. But the downside is a long (and maybe growing) list of parameters in a constructor.
> 
> I personally do have no preference. This new class DicomServiceDependencies feels like a kind of mixture of both approaches. I would also be fine with moving those dependening services to property injection.

I'm personally not a fan of property injection, because the constructor is a sort of "promise" that when you pass it the correct parameters, the class will be usable. With property injection, you break this promise.

Actually I believe the design of DicomServer and DicomServices would benefit from "composition over inheritance", where consumer code does not need to inherit from our DicomService, but instead would somehow "wrap" it and use methods on it instead of inheriting from it. This would have avoided this issue altogether, and also plays a lot more nicely with dependency injection in general. But that would be too much of a breaking change in the context of simple memory optimizations, so I chose the solution that has the smallest possible impact and sets us up in the future to allow further new dependencies in DicomService and DicomServer.

**Comment by amoerie** [comment]
It's benchmark time! @gofal You wanted me to take a look if migrating from ArrayPool to MemoryProvider incurred any costs. These are the results.



**Before**
_______

When running the ServerBenchmarks on the development branch, you get this:

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.22000
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.101
  [Host]     : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT
  Job-ZJXJGO : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT

InvocationCount=128  MaxIterationCount=25  MaxWarmupIterationCount=10  
UnrollFactor=16  

```
|            Method |     Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |    Gen 2 |  Allocated |
|------------------ |---------:|----------:|----------:|---------:|---------:|---------:|-----------:|
|  SendEchoToServer | 1.421 ms | 0.1897 ms | 0.2329 ms |  15.6250 |        - |        - |   97.08 KB |
| SendStoreToServer | 3.981 ms | 0.3187 ms | 0.4031 ms | 234.3750 | 210.9375 | 210.9375 | 3223.56 KB |







**After**
_______

When running the ServerBenchmarks on this branch, you get this: 

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.22000
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.101
  [Host]     : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT
  Job-JZJRQP : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT

InvocationCount=128  MaxIterationCount=25  MaxWarmupIterationCount=10  
UnrollFactor=16  

```
|            Method |     Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |---------:|----------:|----------:|--------:|------:|------:|----------:|
|  SendEchoToServer | 1.035 ms | 0.0321 ms | 0.0428 ms | 15.6250 |     - |     - | 100.23 KB |
| SendStoreToServer | 2.921 ms | 0.1014 ms | 0.1246 ms | 23.4375 |     - |     - | 151.25 KB |


So, here are some of my conclusions:

- Our DicomClient + DicomServer seems to incur a memory allocation cost of around 100kb per request sent + received in the same process. Compared to the ArrayPool based implementation, we seem to have increased this cost by a few kB, but I'm not entirely sure this all comes from the new MemoryProvider.
- The sending of a C-STORE request for a file smaller than 1 MB now allocates 95% less memory compared to development (depending on the file size of course, this benchmark was run with a 500kb file)
- Both C-ECHO and C-STORE are now considerably faster on average.

**Comment by amoerie** [comment]
For sport, these are benchmarks for sending an 11MB DICOM file (Mammo)

**Before**
-----------


``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.22000
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.101
  [Host]     : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT
  Job-VEQQYQ : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT

InvocationCount=128  MaxIterationCount=25  MaxWarmupIterationCount=10  
UnrollFactor=16  

```
|            Method |      Mean |     Error |    StdDev |     Gen 0 |     Gen 1 |     Gen 2 |   Allocated |
|------------------ |----------:|----------:|----------:|----------:|----------:|----------:|------------:|
| SendStoreToServer | 21.949 ms | 1.3789 ms | 1.8408 ms | 1117.1875 | 1070.3125 | 1062.5000 | 46918.83 KB |


**After**
-----------

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.22000
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.101
  [Host]     : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT
  Job-XNQFXO : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT

InvocationCount=128  MaxIterationCount=25  MaxWarmupIterationCount=10  
UnrollFactor=16  

```
|            Method |      Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |----------:|----------:|----------:|--------:|------:|------:|----------:|
| SendStoreToServer | 20.377 ms | 2.9066 ms | 3.8802 ms | 46.8750 |     - |     - | 321.19 KB |

So yay, we don't even allocate 1MB! We finally have full streaming of a DICOM file from client to server without ever loading the full DICOM file in memory. 🎉

File: Tests/FO-DICOM.Benchmark/Program.cs (hunk 2 above)
Comment (amoerie) [reply Tests/FO-DICOM.Benchmark/Program.cs]: Reverted!

**Review by gofal — APPROVED** [review]
👍
